lib/kino/input.ex

```diff
@@ -0,0 +1,228 @@
+
+      name = Kino.Input.read(input)
+  """
+
+  defstruct [:attrs]
+
+  @type t :: %__MODULE__{attrs: map()}
+
+  defp new(attrs) do
+    token = Kino.Bridge.generate_token()
+    id = {token, attrs} |> :erlang.phash2() |> Integer.to_string()
+    attrs = Map.put(attrs, :id, id)
+    %__MODULE__{attrs: attrs}
+  end
+
+  @doc """
+  Creates a new text input.
+
+  ## Options
+
+    * `:default` - the initial input value. Defaults to `""`
+
+    * `:multiline` - whether the input should be multiline
+      text box. Defaults to `false`
+  """
+  @spec text(String.t(), keyword()) :: t()
+  def text(label, opts \\ []) when is_binary(label) and is_list(opts) do
+    default = Keyword.get(opts, :default, "")
+    multiline = Keyword.get(opts, :multiline, false)
+    new(%{type: :text, label: label, default: default, multiline: multiline})
+  end
+
+  @doc """
+  Creates a new password input.
+
+  This is similar to text input, except the content is not
+  visible by default.
+
+  ## Options
+
+    * `:default` - the initial input value. Defaults to `""`
+  """
+  @spec password(String.t(), keyword()) :: t()
+  def password(label, opts \\ []) do
+    default = Keyword.get(opts, :default, "")
+    new(%{type: :password, label: label, default: default})
+  end
+
+  @doc """
+  Creates a new number input.
+
+  The input value is can be either a float or `nil`.
+
+  ## Options
+
+    * `:default` - the initial input value. Defaults to `nil`
+  """
+  @spec number(String.t(), keyword()) :: t()
+  def number(label, opts \\ []) when is_binary(label) and is_list(opts) do
+    default = Keyword.get(opts, :default, nil)
+    new(%{type: :number, label: label, default: default})
+  end
+
+  @doc """
+  Creates a new URL input.
+
+  The input value can be either a valid URL string or `nil`.
+
+  ## Options
+
+    * `:default` - the initial input value. Defaults to `nil`
+  """
+  @spec url(String.t(), keyword()) :: t()
+  def url(label, opts \\ []) when is_binary(label) and is_list(opts) do
+    default = Keyword.get(opts, :default, nil)
+    new(%{type: :url, label: label, default: default})
+  end
+
+  @doc """
+  Creates a new select input.
+
+  The input expects a list of options in the form `[{value, label}]`,
+  where `value` is an arbitrary term and `label` is a descriptive
+  string.
+
+  ## Options
+
+    * `:default` - the initial input value. Defaults to the first
+      value from the given list of options
+
+  ## Examples
+
+      Kino.Input.select("Language", [en: "English", fr: "Français"])
+
+      Kino.Input.select("Language", [{1, "One"}, {2, "Two"}, {3, "Three"}])
+  """
+  @spec select(String.t(), list({value :: term(), label :: String.t()}), keyword()) :: t()
+  def select(label, options, opts \\ [])
+      when is_binary(label) and is_list(options) and is_list(opts) do
+    if options == [] do
+      raise ArgumentError, "expected at least on option, got: []"
+    end
+
+    values = Enum.map(options, &elem(&1, 0))
```

Comment:
We should probably normalize all labels and values to strings here.

Reply:
Labels yeah, but values can be anything no?

Reply:
The problem is that the browser is going to send those as strings and doing the mapping will be annoying. Alternatively we support rendering only lists of values, such that: `select("Category", ~w(Books Cars ...))` and avoid this problem for now. Over the wire we still send it as a list of tuples though.

Reply:
On Livebook side we use indices for the select and then map to the corresponding key, see [here](https://github.com/livebook-dev/livebook/blob/041764096c97414bc896a5067a3a5cd3901cb4ca/lib/livebook_web/live/output/input_component.ex#L51-L55).

Reply:
Ah, smart! So ignore me :)